text/0113-secure-script-container.md

```diff
@@ -1,88 +1,121 @@
 
-* Run or include the third party library script inside this sandbox
-  * This allows the code to run in a non-patched environment, while the rest of the document is running synthetic shadow
-* Package the above functionality in a web component which semantically matches the `<script>` tag.
+* Create a JS sandbox before the Synthetic Shadow Polyfill is evaluated.
+  * The sandbox is created using a same-domain iframe.
 
+* Use this JS sandbox to virtualize the DOM APIs in order to prevent these APIs to be subject to the Synthetic Shadow Polyfill patches:
+  * This can be done by caching references to the to-be-patched APIs inside the sandbox before the synthetic shadow is evaluated.
 
-### Native Shadow
+* Evaluate one or more third party libraries inside this sandbox
+  * This allows the code to run in a non-patched environment, while the rest of the document is running synthetic shadow.
 
-For applications which are using the native shadow, the underlying implementation will be different. There are two approaches we can choose from:
+* Package the above functionality in a web component which semantically matches the `<script>` tag API.
 
-1. Virtualize every shadow DOM tree in the outer window into a single DOM structure which this container would see as the default DOM.
-2. Implement a `document.querySelector` API which traverses the shadow DOM naturally.
+* Provide an additional mechanism to remap globals when needed.
+    * The sandbox does NOT include any global name defined in the main window, whether it is defined before or after the creation of the sandbox. But the `<virtual-script>` can opt-in to share a global value between the sandbox and the main window, this must be done via the `extra-globals` argument passed to the `<virtual-script extra-globals="dataLayer">` element.
+
+Other considerations for Phase I:
+
+ * due to the explicit order of evaluation required for this to function, we plan to only support one sandbox that can be shared by multiple virtual scripts.
+ * budget is around 5kb (minify/gzip).
+ * performance penalties will be added to the virtual scripts as they run inside a membrane:
+   * evaluation budget: 30% slower than evaluating the same script in the main window.
+   * operation budget: 50% slower than executing the same task is the main window.
+   * 3rd party scripts are rarely part of the critical path for the application to function.
+
+### Phase II: Native Shadow
+
+For applications which are using the native shadow, the underlying implementation will be different, a more complicated one. We are currently proposing the following characteristics for the virtual-script design:
+
+* During the evaluation of the virtual script component class, patch `Element.prototype.attachShadow`, to record any shadow created in the app.
+
+* When evaluating code via `<virtual-script>` element, create a new sandbox with patched DOM APIs to simulate a flatten tree.
```

Comment:
Why are we proposing to flatten instead of preserving the original DOM's hierarchy?

> This will require a query selector mechanism that can translate regular selectors into cross shadow selections

Eg: how will we reason with matching `element1 > element2` and *not* matching `element1 + element2`

Reply:
We are talking about the native shadow here, so, there is no DOM hierarchy here that can work, we have to simulate it. I'm not sure I understand the question, but the point is, that in order for those 2 queries to function, specially when element1 is in one shadow, and element2 is in another shadow... how can you make that work? well, you have to reconstruct in memory a flatten representation of the intersection of those two shadow fragments, and do a query based on that... obviously that will be too slow, too complicated, but we can do something that represents that without having to construct it in memory. I have some ideas.

Reply:
Ok. When we say a flattened tree to me sounds like we would take something like this:

```html
<x-app>
  ▼ #shadow-root
    <x-child>
      ▼ #shadow-root
        <button />
    </x-child>
</x-app>
```

and flatten it to something like this:

```html
<x-app />
<x-child />
```

Reply:
no, flatten tree reference to how browsers compute the UI elements, in this case, the flatten tree is:

```
x-app
   x-child
      button
```